docs/in-depth-guide-scraper-functions.md

```diff
@@ -0,0 +1,319 @@
+
+The language code is based on BCP 47 standards.
+For a comprehensive list of BCP 47 language codes, refer to this GitHub Gist:
+https://gist.github.com/typpo/b2b828a35e683b9bf8db91b5404f1bd1
+
+```py
+>>> scraper.language()
+'en'
+```
+
+### `links() -> List[Dict[str, str]]`
+
+Returns a `list` of all links found in the page HTML defined in an `<a>` element. For each link, the attributes of the HTML element are returned as a `dict`.
+
+```py
+>>> scraper.links()
+[
+	{
+		'class': ['popup-toggle-button'],
+		'aria-label': 'Toggle main-navigation popup',
+  		'aria-haspopup': 'true',
+  		'href': '#main-navigation-popup'
+	},
+	... # etc.
+]
+```
+
+### `site_name() -> str | None`
+
+Returns the website name, as defined in the page's HTML. If the page does not define this, this function returns None
+
+```py
+>>> scraper.site_name()
+None
+```
+
+### `to_json() -> List[str, str]`
+
+Returns the output of all functions implemented by this scraper as a `dict`.
+
+```py
+>>> scraper.to_json()
+{
+	'author': 'Good Food team',
+	'canonical_url': 'https://www.bbcgoodfood.com/recipes/monster-cupcakes',
+	'category': 'Treat',
+	... # etc.
+}
+```
+
+
+
+## Optional methods
+
+### `category() -> str`
+
+Returns the category of the recipe.
+
+```py
+>>> scraper.category()
+'Treat'
+```
+
+### `cook_time() -> int`
+
+Returns the time to cook the recipe in minutes, excluding any time to prepare ingredients.
+
+```py
+>>> scraper.cook_time()
+20
+```
+
+### `cuisine() -> str`
+
+Returns the cuisine of the recipe.
+
+```py
+>>> scraper.cuisine()
+# Not implemented!
+```
+
+### `nutrients() -> Dict[str, str]`
+
+Returns a `dict` of nutrition information. Each nutrition item is usually given per unit of yield, e.g. per servings, per item. The keys of the `dict` are the nutrients (including calories) and the values are the amount of that nutrient, including the unit.
+
+```py
+>>> scraper.nutrients()
+{
+	'calories': '389 calories',
+ 	'fatContent': '19 grams fat',
+ 	'saturatedFatContent': '9 grams saturated fat',
+ 	'carbohydrateContent': '53 grams carbohydrates',
+ 	'sugarContent': '36 grams sugar',
+ 	'fiberContent': '1 grams fiber',
+ 	'proteinContent': '5 grams protein',
+ 	'sodiumContent': '0.3 milligram of sodium'
+}
+```
+
+### `prep_time() -> int`
+
+Returns the time to prepare the ingredients for the recipe in minutes.
+
+```py
+>>> scraper.prep_time()
+30
+```
+
+### `ratings() -> float`
+
+Returns the recipe rating. Where this is available, this is usually the average of all the ratings given to the recipe on the website.
+
+```py
+scraper.ratings()
+# Not implemented!
+```
+
+### `reviews() -> List[Dict[str, str]]`
+
+Returns a `list` of reviews about the recipe from the website. Each review is a `dict` containing the reviewer's name (`str`) and their review (`str`).
+
+> [!WARNING]
+>
+> As far as I can tell, only the `AllRecipes` scraper actually implements this. Should we encourage more scrapers to use it, or should we remove the feature?
```

Comment:
Nice catch :+1: 

I wonder.. I think we could check how many scrapers _might_ support [`schema.org` reviews](https://schema.org/Review) if we added it as a schema field.  That could take us from one to many fairly quickly.

I'll try to do that analysis soon (note: if I haven't responded to this within a day, then it probably means I've forgotten about it).

Reply:
> I wonder.. I think we could check how many scrapers _might_ support [`schema.org` reviews](https://schema.org/Review) if we added it as a schema field. That could take us from one to many fairly quickly.

I had an idea for a way to run a naive, basic analysis.  Based on the commands below, 25 of our scrapers look like they contain `Review` objects as structured data:

```shell
$ grep -rwl '@type.*"Review"' tests/test_data | wc -l
25
```

...so maybe 10% of scrpers _could_ return `reviews` if we implement that there?  That would seem valuable enough to open a feature request for.

Reply:
If we can extract this from the schema, that I think this is would be worth doing. :+1:

Reply:
Great - I've filed #870 to track that, and will take a look into more of the details soon.